roles/manage_agents/tasks/add_agents.yml

```diff
@@ -0,0 +1,82 @@
+    # the selection of new agents is serialized by resource class
+    - name: Acquire agent lock
+      kubernetes.core.k8s:
+        state: present
+        apply: true
+        server_side_apply:
+          field_manager: "{{ cluster_order_holder_id }}"
+        definition:
+          apiVersion: coordination.k8s.io/v1
+          kind: Lease
+          metadata:
+            name: "agent-{{ manage_agents_resource_class }}-lock"
+            namespace: "{{ default_agent_namespace }}"
+          spec:
+            holderIdentity: "{{ cluster_order_holder_id }}"
+      register: manage_agents_lease
+      until: manage_agents_lease is successful
+      retries: 20
+      delay: 5
+
+    - name: Retrieve the list of available agents with a matching machine resource class
+      kubernetes.core.k8s_info:
+        kind: Agent
+        api_version: agent-install.openshift.io/v1beta1
+        namespace: "{{ default_agent_namespace }}"
+        label_selectors:
+          - "!{{ cluster_order_label }}"
+          - "{{ esi_agent_resource_class_label }}={{ manage_agents_resource_class }}"
+      register: manage_agents_available
+
+    # Agents may be picked-up outside of the fulfillment API, so we need to exclude them
+    - name: Filter out agents already allocated to a cluster
+      ansible.builtin.set_fact:
+        manage_agents_available: >
+          {{ manage_agents_available.resources
+          | selectattr('spec.clusterDeploymentName.name', 'undefined') }}
+
+    - name: Count the number of agents to add to the cluster
+      ansible.builtin.set_fact:
+        manage_agents_selected_count: >
+          {{ manage_agents_desired_count | int - manage_agents_allocated.resources | length }}
+
+    - name: Add selected agents to the cluster
+      kubernetes.core.k8s_json_patch:
+        api_version: agent-install.openshift.io/v1beta1
+        kind: Agent
+        name: "{{ item.metadata.name }}"
+        namespace: "{{ item.metadata.namespace }}"
+        patch:
+          - op: add
+            path: /metadata/labels/{{ cluster_order_label | cloudkit.service.json_pointer_escape }}
+            value: "{{ manage_agents_cluster_order_name }}"
+      loop: "{{ manage_agents_available[: (manage_agents_selected_count | int)] }}"
+      loop_control:
+        label: "Adding agent {{ item.metadata.name }}"
+      register: manage_agents_added
+
+    - name: Count the number of agents effectively added to the cluster
+      ansible.builtin.set_fact:
+        manage_agents_added_count: >
+          {{ manage_agents_added.results | selectattr('changed') | list | length }}
+
+    - name: Print a message if not all agents were added
+      ansible.builtin.debug:
+        msg: "{{ manage_agents_added_count }} agents were added to the cluster instead of {{ manage_agents_selected_count }}. Please check agents availability."
+      when: manage_agents_added_count | int != manage_agents_selected_count | int
+
+  always:
+
+    - name: Delete agent lock
```

Comment:
@mhrivnak pointed out in #38 that a failure here will leave stale leases floating around. If we include the ansible job id in the lease, we can the implement a cleanup job that looks for leases, checks if the job is active or not, and cleans them if necessary.

In #38 I have the job id encoded in the `holderIdentity`, but it probably makes more sense to add it as a label; maybe `cloudkit.openshift.io/aap-job-id`?

I'll make a new pr that applies that change to both places.